DDP + C10D sparse all_reduce changes

H-Huang opened the pull request:
Summary:
## Changes

prototyping sparse allreduce using the sparse dispatch key. When passing in sparse tensors into `dist.allreduce()` we can execute our dispatched function.

prior to this change, passing a sparse tensor into `allreduce()` will error out with `Tensor must be dense...`

## Example script

```python
# python -m torch.distributed.run --nnodes=1 --nproc_per_node=2 this_script.py

import torch
import torch.distributed as dist

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    a = torch.tensor([[0, 2.], [3, 0]]).to(rank)
    a = a.to_sparse()
    print(f"rank {rank} - a: {a}")
    dist.all_reduce(a)

if __name__ == "__main__":
    main()
```

output:
```
rank 1 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:1', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
rank 0 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:0', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
```

Test Plan:
Testing commands (OSS):

```
# python
pytest test/distributed/test_c10d_nccl.py -vsk test_sparse_allreduce_ops

# c++
build/bin/ProcessGroupNCCLTest --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Testing commands (internal, ondemand GPU):
ddp tests:
```
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output

# Get the .par file from the previous command and use it below
TORCH_SHOW_CPP_STACKTRACE=1 /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_ddp_set_sparse_metadata
```

c10d tests:
```
# build tests and run with log output (python)
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output
NCCL_DEBUG=WARN /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_sparse_allreduce_ops

# python
NCCL_DEBUG=WARN buck test mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d -- --exact 'caffe2/test/distributed:c10d - test_sparse_allreduce_ops (test_c10d_nccl.ProcessGroupNCCLTest)'

# c++
NCCL_DEBUG=WARN buck run mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/cpp/c10d:ProcessGroupNCCLTest -- --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Differential Revision: D46724856

Pulled By: H-Huang

H-Huang commented on torch/nn/parallel/distributed.py:
```diff
@@ -1234,8 +1234,8 @@ def _assign_modules_buffers(self):
-        if dist.get_debug_level() == dist.DebugLevel.OFF:
-            return {}
+        # if dist.get_debug_level() == dist.DebugLevel.OFF:
+        #     return {}
```
@rohan-varma, is it okay to remove this check? Param to name mapping is needed for sparse allreduce when global unique id is passed in. I assume cost to create this mapping is negligible?

rohan-varma replied:
yep

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -305,6 +305,30 @@ def broadcast(xs, rootRank, rootTensor):
+
+        # sparse allreduce call is wrapped in a try catch since the c10d API is only available in the nccl experimental branch
+        try:
+            work = pg.allreduce([sparse_tensor])
```
do we want to try with `get_future()` API?

rohan-varma commented on test/cpp/c10d/ProcessGroupNCCLTest.cpp:
```diff
@@ -731,3 +878,16 @@ TEST_F(ProcessGroupNCCLTest, testBackendName) {
+  {
+    TemporaryFile file;
+    testSparseAllreduce(file.path, rank_, size_);
+    // testSparseAllreduceLarge(file.path, rank_, size_);
```
does the test work?

H-Huang replied:
yes, both work and operate on different sizes. Will uncomment it

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3049,6 +3073,74 @@ def test_new_group_local_sync_duplicated_pg(self):
+
+        def forward(self, inputs):
+            embedded = self.embedding(inputs)
+            # embedded shape: (batch_size, sequence_length, embedding_dim)
```
nit: remove comments before landing

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3049,6 +3073,74 @@ def test_new_group_local_sync_duplicated_pg(self):
+            loss = output.sum()
+
+            # backward pass
+            loss.backward()
```
can we validate equivalence somehow? i.e. gradients are the same with and without sparse allreduce / setting sparse metadata?

rohan-varma commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -12,6 +12,12 @@
 
+#ifndef NCCL_MAJOR
+  #ifndef NCCL_MINOR
+    #define NCCL_NOT_DEFINED
```
where is it used?

H-Huang replied:
not used anymore, cleaning it up, thanks!

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1556,8 +1557,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
       if (!avoidRecordStreams_) {
-        c10::cuda::CUDACachingAllocator::recordStream(
-            inputs[i].storage().data_ptr(), ncclStream);
+        // TODO: fix for sparse tensors
```
don't we need this fixed, since we need to support sparse tensors here?

H-Huang replied:
old TODO that isn't relevant anymore

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1556,8 +1557,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
+        } else {
+          c10::cuda::CUDACachingAllocator::recordStream(
+              inputs[i].values().storage().data_ptr(), ncclStream);
+          c10::cuda::CUDACachingAllocator::recordStream(
```
nit: I think the lack of this was a bug that stumped us for a while, so can you add a comment why this is needed? 

IIRC - since the tensor will be used on a different (the nccl) stream, we need to call recordStream to ensure the tensor is not prematurely freed?

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1816,6 +1827,86 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
       profilingTitle);
 }
 
+// Sparse allreduce (note: ProcessGroupGloo.cpp::AsyncSparseAllreduceWork)
```
what is the `note:` for?

H-Huang replied:
thanks, not needed

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1816,6 +1827,86 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+        auto recvIndices = indices[0] * colSize;
+
+        // prevent output and recvIndices from being freed
+        c10::cuda::CUDACachingAllocator::recordStream(
```
nit: not needed to change in this PR, but theoretically we can just record `output` for every `collective` in the `collective` implementation?

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1816,6 +1827,86 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+                tensors[i].sizes());
+          } else {
+            tensors[i] = outputTensors[i].to_sparse();
+            // int num_rows = tensors[i].numel() / tensors[i].sizes()[1];
```
remove?

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1816,6 +1827,86 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+        }
+      },
+      OpType::ALLREDUCE,
+      "nccl:all_reduce");
```
maybe make this `nccl:all_reduce_sparse` for the profiler to distinguish

H-Huang replied:
nice catch, thanks!

rohan-varma commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -25,13 +25,15 @@ class TORCH_API GradBucket {
       std::vector<size_t> offsets,
       std::vector<size_t> lengths,
       std::vector<c10::IntArrayRef> sizes_vec,
+      c10::optional<at::Tensor>& global_unique_id,
```
Will defer to you, but `global_unique_id` seems like a higher level implementation detail that doesn't fully reflect how users can use this variable. Can we name it something more conceptual / easier to understand?

rohan-varma commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -70,6 +72,10 @@ class TORCH_API GradBucket {
     return index_ == bucket_count_ - 1;
   }
 
+  c10::optional<at::Tensor>& getGlobalUniqueId() {
```
pls check if it needs a lock

rohan-varma commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -46,7 +46,14 @@ c10::intrusive_ptr<c10::ivalue::Future> FP16CompressCommHook::runHook(
+  // sparse gradient indices update case
+  AllreduceOptions opts = AllreduceOptions();
+  opts.sparseIndices = bucket.getGlobalUniqueId().value();
+  return state_->allreduce_sparse(tensors, opts)->getFuture();
```
i might be confused, but does this mean we always call sparse allreduce if IS_NCCL_EXP = true (even if we have a dense tensor)? 

Have we tested the case where IS_NCCL_EXP=true and we're allreducing a dense tensor?

also I'm confused on this - if we have the dispatcher, why do we need to call `allreduce_sparse` explicitly here, shouldn't we call just `allreduce` and the dispatcher figures out which implementation to call?

H-Huang replied:
This is a good catch. I'm updating the code to only call into allreduce and utilize the dispatcher

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -438,6 +438,25 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
+    // Copy the indices of sparse metadata
+    if (sparse_metadata_) {
+      grad = grad.coalesce();
+      REDUCER_CHECK(
```
since we're turning this on by default now, the check can probably be moved to the Reducer constructor. Here, it would probably be called on every parameter every iteration.

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -438,6 +438,25 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
+          param_names_.size() != 0, logger_, "No parameter names were found");
+      std::string& param_name = param_names_[variable_index];
+      REDUCER_CHECK(
+          (*sparse_metadata_).count(param_name) > 0,
```
nit: we also access the map on L452, maybe read it only once and save it

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -462,6 +481,7 @@ std::vector<c10d::GradBucket> Reducer::get_grad_buckets(
   for (const auto i : c10::irange(buckets_.size())) {
     auto& bucket = buckets_[i];
     auto variables_for_bucket = get_variables_for_bucket(i, bucket);
+    c10::optional<at::Tensor> unused = c10::nullopt;
```
what's this for?

H-Huang replied:
updated so it isnt needed anymore

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -470,6 +490,7 @@ std::vector<c10d::GradBucket> Reducer::get_grad_buckets(
         bucket.offsets,
         bucket.lengths,
         bucket.sizes_vec,
+        unused,
```
what argument is this? can we just default it?

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1538,7 +1560,21 @@ void Reducer::finalize_backward() {
         : comm_hook_->parseHookResult(bucket.future_work->value());
     if (bucket.expect_sparse_gradient) {
-      bucket.gradients.copy_(future_result);
+      if (sparse_metadata_ && bucket.global_unique_id.has_value()) {
```
is it possible for sparse_metadata = true and bucket.global_unique_id.has_value = False? In what cases can that happen?

H-Huang replied:
there shouldn't be a case, so yeah the check seems redundant

rohan-varma commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -500,6 +503,10 @@ class TORCH_API Reducer {
   std::unique_ptr<CommHookInterface> comm_hook_;
+
+  // sparse metadata
+  std::unique_ptr<std::map<std::string, at::Tensor>> sparse_metadata_;
```
nit: I would add a description of what the key here is and what the value is conceptually (i.e. I feel like `global_unique_id` is an unfamiliar term to most users). The API is private for now but might be good for conceptual undestanding.

rohan-varma reviewed (approved):
LGTM, thanks for adding this! Have a few comments inline. Main asks: 

1. Can we add more comprehensive unittesting - such as DDP + set_sparse_metadata + sparse allreduce, and verify (up to a tolerenance) with DDP + the same tensor, but dense? 
2. What happens now if I do DDP with sparse grads + PG NCCL, but don't call set_sparse_metadata? Are we supporting this feature, and if so are we sure about the correctness? If we're not sure about correctness we should probably just error out for now, if the scope of the PR is to just support with set_sparse_metadata. If adding tests for this is not in scope for this PR, let's file a follow up issue 
3. Consider renaming `global_unique_id` to something more conceptual and adding a docstr on it
4. I don't completely understand why we need to change the allreduce in default_comm_hooks.cpp and case it on USE_NCCL_EXP. Could you clarify what happens if we end up in that codepath, USE_NCCL_EXP=true, but we're allreducing a dense tensor? And why we need to distinguish there btwn allreduce_sparse and allreduce_, instead of using the dispatcher.

H-Huang committed:
DDP + C10D sparse all_reduce changes (#103916)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103916

## Changes

prototyping sparse allreduce using the sparse dispatch key. When passing in sparse tensors into `dist.allreduce()` we can execute our dispatched function.

prior to this change, passing a sparse tensor into `allreduce()` will error out with `Tensor must be dense...`

## Example script

```python
# python -m torch.distributed.run --nnodes=1 --nproc_per_node=2 this_script.py

import torch
import torch.distributed as dist

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    a = torch.tensor([[0, 2.], [3, 0]]).to(rank)
    a = a.to_sparse()
    print(f"rank {rank} - a: {a}")
    dist.all_reduce(a)

if __name__ == "__main__":
    main()
```

output:
```
rank 1 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:1', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
rank 0 - a: tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([2., 3.]),
       device='cuda:0', size=(2, 2), nnz=2, layout=torch.sparse_coo)
allreduce_sparse_cuda_
tensor.is_sparse() = 1
in ProcessGroupNCCL::allreduceSparse
```

Test Plan:
Testing commands (OSS):

```
# python
pytest test/distributed/test_c10d_nccl.py -vsk test_sparse_allreduce_ops

# c++
build/bin/ProcessGroupNCCLTest --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Testing commands (internal, ondemand GPU):
ddp tests:
```
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output

# Get the .par file from the previous command and use it below
TORCH_SHOW_CPP_STACKTRACE=1 /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_ddp_set_sparse_metadata
```

c10d tests:
```
# build tests and run with log output (python)
buck build mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d --show-full-output
NCCL_DEBUG=WARN /data/sandcastle/boxes/fbsource/buck-out/v2/gen/fbcode/c8344b52091f4f7f/caffe2/test/distributed/__c10d__/c10d.par -r test_sparse_allreduce_ops

# python
NCCL_DEBUG=WARN buck test mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/distributed:c10d -- --exact 'caffe2/test/distributed:c10d - test_sparse_allreduce_ops (test_c10d_nccl.ProcessGroupNCCLTest)'

# c++
NCCL_DEBUG=WARN buck run mode/opt -c hpc_comms.use_nccl=exp //caffe2/test/cpp/c10d:ProcessGroupNCCLTest -- --gtest_filter=ProcessGroupNCCLTest.testSparseAllreduce
```

Reviewed By: rohan-varma

Differential Revision: D46724856

Pulled By: H-Huang

fbshipit-source-id: 5f11fcd18699e230bac9b43f358b26069c0ff122
